core/modules/qdisp/QueryRequest.cc

```diff
@@ -213,6 +252,15 @@ bool QueryRequest::isCancelled() {
 }
 
 void QueryRequest::cleanup() {
+    LOGF_DEBUG("%1% QueryRequest::cleanup()" %_jobId);
+    {
+        std::lock_guard<std::mutex> lock(_finishStatusMutex);
+        if (_finishStatus == ACTIVE) {
+            LOGF_ERROR("%1% QueryRequest::cleanup called before finished" % _jobId);
+            return;
+        }
+    }
+    // These need to be outside the mutex lock, or you could delete the mutex before it is freed.
```

Comment:
I'd +1 Andy's recommend here -- it seems a lot less cumbersome to me?  Maybe prefix with a one-line comment to clarify:
```C++
// Vacate both _jobQuery and _keepAlive before freeing either
```